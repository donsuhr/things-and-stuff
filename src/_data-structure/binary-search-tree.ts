import { TreeNode as Node } from './tree-node';

export class BinarySearchTree<T> {
    #root: Node<T> | null = null;

    // remove(): no implementation, complicated

    insert(val: T) {
        const node = new Node<T>(val);

        if (this.#root === null) {
            this.#root = node;
        } else {
            let pointer = this.#root;
            while (pointer) {
                if (node.value! > pointer.value!) {
                    if (pointer.right === null) {
                        pointer.right = node;
                        break;
                    }
                    pointer = pointer.right;
                } else {
                    if (pointer.left === null) {
                        pointer.left = node;
                        break;
                    }
                    pointer = pointer.left;
                }
            }
        }
    }

    find(val: T) {
        let pointer = this.#root;
        while (pointer !== null) {
            if (pointer.value === val) {
                return pointer;
            }
            if (pointer.value && val > pointer.value) {
                pointer = pointer.right;
            } else {
                pointer = pointer.left;
            }
        }
        return undefined;
    }

    /*
      Depth First Search

      less memory than BFS. However its recursive call stacks can get deep
      does a path exist, solve a maze
      can get slow, dont use on big trees with rare results

      - in order: return a sorted array from bst
      - pre order: useful for copy a tree
      - post order: ?
    */

    depthFirstSearch(
        node = this.#root,
        result: T[] = [],
        order: 'preOrder' | 'inOrder' | 'postOrder' = 'inOrder',
    ) {
        if (order === 'preOrder' && node?.value) {
            result.push(node.value);
        }

        if (node?.left) {
            this.depthFirstSearch(node.left, result, order);
        }

        if (order === 'inOrder' && node?.value) {
            result.push(node.value);
        }

        if (node?.right) {
            this.depthFirstSearch(node.right, result, order);
        }

        if (order === 'postOrder' && node?.value) {
            result.push(node.value);
        }
        return result;
    }

    /*
       Breadth First Search

       use to find the shortest path, most related items, closest friends
       uses more memory
        - avoid on wide tree

       Dijkstraâ€™s : doesn't work with negative numbers
       Bellman-Ford
    */

    breadthFirstSearch() {
        let node = this.#root;
        const queue = [node];
        const result = [];
        let currentCount = 1; // for pretty print
        let nextCount = 0; // for pretty print
        let resultString = '';
        while (queue.length) {
            node = queue.shift()!;
            result.push(node.value);
            currentCount -= 1;
            resultString += `${node.value} `;
            if (node.left) {
                queue.push(node.left);
                nextCount++;
            }
            if (node.right) {
                queue.push(node.right);
                nextCount++;
            }
            if (currentCount === 0) {
                resultString += `\n`;
                [currentCount, nextCount] = [nextCount, currentCount];
            }
        }
        return { result, formatted: resultString };
    }
}
