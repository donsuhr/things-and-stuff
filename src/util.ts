function randomNumber(min = 1, max = 10): number {
    return Math.floor(Math.random() * max) + min;
}

function randomNumbers(n: number): number[] {
    const ret: number[] = [];
    for (let i = 0; i < n; i += 1) {
        ret.push(Math.floor(Math.random() * 100));
    }
    return ret;
}

function isSorted(A: any[], reverse = false): boolean {
    for (let i = 0; i < A.length; i += 1) {
        if (reverse && A[i] < A[i + 1]) {
            return false;
        }
        if (!reverse && A[i] > A[i + 1]) {
            return false;
        }
    }
    return true;
}

export { randomNumbers, isSorted, randomNumber };
