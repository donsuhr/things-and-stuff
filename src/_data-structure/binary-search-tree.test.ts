import { BinarySearchTree } from './binary-search-tree';
// import { BinarySearchTreeDaily } from '../_daily/binarySearchTree';
import { isSorted } from '../util';

describe.each([
    { fn: BinarySearchTree },
    // { fn: BinarySearchTreeDaily }
])('Binary Search Tree: $fn.name', ({ fn: CurrentBinarySearchTree }) => {
    let sut: BinarySearchTree<number>;

    beforeEach(() => {
        //     9
        //  4     20
        // 1  6  15 170

        sut = new CurrentBinarySearchTree<number>();
        sut.insert(9);
        sut.insert(4);
        sut.insert(6);
        sut.insert(20);
        sut.insert(170);
        sut.insert(15);
        sut.insert(1);
    });

    it('creates a tree', () => {
        const dfs = sut.depthFirstSearch();
        expect(sut.depthFirstSearch()).toEqual(dfs);
        expect(isSorted(dfs)).toBe(true);
    });

    it('breadthFirstSearch', () => {
        expect(sut.breadthFirstSearch().result).toEqual([
            9, 4, 20, 1, 6, 15, 170,
        ]);
        expect(sut.breadthFirstSearch().formatted).toMatchSnapshot();
    });

    it('finds existing', () => {
        const found = sut.find(20)!;
        expect(found.value).toBe(20);
        expect(found.left!.value).toBe(15);
    });

    it('does not find missing', () => {
        expect(sut.find(44)).not.toBeDefined();
    });

    describe('Depth First Search', () => {
        it('preOrder', () => {
            expect(
                sut.depthFirstSearch(undefined, undefined, 'preOrder'),
            ).toEqual([9, 4, 1, 6, 20, 15, 170]);
        });

        it('inOrder', () => {
            expect(
                sut.depthFirstSearch(undefined, undefined, 'inOrder'),
            ).toEqual([1, 4, 6, 9, 15, 20, 170]);
        });

        it('postOrder', () => {
            expect(
                sut.depthFirstSearch(undefined, undefined, 'postOrder'),
            ).toEqual([1, 6, 4, 15, 170, 20, 9]);
        });
    });
});
