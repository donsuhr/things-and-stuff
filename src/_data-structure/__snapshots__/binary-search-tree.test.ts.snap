// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Binary Search Tree: BinarySearchTree breadthFirstSearch 1`] = `
"9 
4 20 
1 6 15 170 
"
`;

exports[`Binary Search Tree: BinarySearchTreeDaily breadthFirstSearch 1`] = `
"9 
4 20 
1 6 15 170 
"
`;
